File: test/Typhon.Protocol.Tests/GoldenSupport.cs
using System;
using System.Globalization;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using NUnit.Framework;
using Typhon.Protocol;

namespace Typhon.Protocol.Tests;

/// <summary>
/// The golden-vector pattern (05-sdks § 4): per case a <c>.bin</c> holding the exact bytes and a <c>.json</c> holding what decoding them must produce, both
/// regenerated only under <c>TYPHON_UPDATE_GOLDEN=1</c>.
/// </summary>
/// <remarks>
/// <para>
/// <b>Why a committed artefact and not a computed expectation.</b> The .NET encoder, the TypeScript decoder and the engine must agree on bytes, and they cannot
/// agree on a value each computes for itself. A committed <c>.bin</c> is the only witness that survives a change to the code that produced it — a
/// self-computed expectation would silently follow the bug. Regenerating by default would defeat the mechanism: the vectors would rewrite themselves to match
/// whatever the code now does, and the test could never fail.
/// </para>
/// <para>
/// <b>Numbers in expectations are IEEE bit patterns</b>, as 16 hex digits (<see cref="Bits"/>): decoded values must be bit-exact across languages (W1), and a
/// JSON number cannot carry NaN, infinities or negative zero, nor promise the last bit survives a round trip through a formatter.
/// </para>
/// <para>
/// The directory is found by walking up to the project file rather than by copying content to the output directory, because regeneration has to write into
/// the source tree that gets committed, not into <c>bin/</c>.
/// </para>
/// </remarks>
internal static class Golden
{
    private const string UpdateVariable = "TYPHON_UPDATE_GOLDEN";

    private static readonly JsonSerializerOptions Indented = new() { WriteIndented = true };

    internal static bool Updating => Environment.GetEnvironmentVariable(UpdateVariable) == "1";

    /// <summary>Asserts <paramref name="bytes"/> and <paramref name="expectation"/> equal the committed vector, or rewrites it on a regeneration run.</summary>
    internal static void Assert(string caseName, byte[] bytes, JsonObject expectation)
    {
        var directory = Directory();
        var binPath = Path.Combine(directory, caseName + ".bin");
        var jsonPath = Path.Combine(directory, caseName + ".json");

        // "\n", not Environment.NewLine: regenerating on Windows and on Linux must produce the same file.
        var json = expectation.ToJsonString(Indented).ReplaceLineEndings("\n") + "\n";

        if (Updating)
        {
            System.IO.Directory.CreateDirectory(directory);
            File.WriteAllBytes(binPath, bytes);
            File.WriteAllText(jsonPath, json, new UTF8Encoding(false));
            return;
        }

        NUnit.Framework.Assert.That(File.Exists(binPath), Is.True, $"missing golden vector {caseName}.bin — regenerate with {UpdateVariable}=1 and commit it");
        var committedBytes = File.ReadAllBytes(binPath);
        var committedJson = File.ReadAllText(jsonPath).ReplaceLineEndings("\n");

        NUnit.Framework.Assert.Multiple(() =>
        {
            NUnit.Framework.Assert.That(Convert.ToHexString(bytes), Is.EqualTo(Convert.ToHexString(committedBytes)), $"{caseName}.bin drifted from the vector");
            NUnit.Framework.Assert.That(json, Is.EqualTo(committedJson), $"{caseName}.json drifted from the vector");
        });
    }

    /// <summary>
    /// A double's IEEE bits as 16 lower-case hex digits, or <c>"nan"</c>: JavaScript cannot observe a NaN's sign or payload, so NaN is compared as a token
    /// while every other value, signed zeros and infinities included, is compared bit for bit.
    /// </summary>
    internal static string Bits(double value) =>
        double.IsNaN(value) ? "nan" : BitConverter.DoubleToInt64Bits(value).ToString("x16", CultureInfo.InvariantCulture);

    internal static JsonArray Bits(ReadOnlySpan<double> values)
    {
        var array = new JsonArray();
        foreach (var v in values)
        {
            array.Add(Bits(v));
        }

        return array;
    }

    internal static string Hex(ReadOnlySpan<byte> bytes) => Convert.ToHexString(bytes).ToLowerInvariant();

    internal static JsonNode CodecJson(CatalogCodec codec) => JsonSerializer.SerializeToNode(codec, CatalogJson);

    private static readonly JsonSerializerOptions CatalogJson = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingDefault,
    };

    private static string Directory()
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        while (directory != null && directory.GetFiles("Typhon.Protocol.Tests.csproj").Length == 0)
        {
            directory = directory.Parent;
        }

        NUnit.Framework.Assert.That(directory, Is.Not.Null, "could not locate the test project directory from the test assembly location");
        return Path.Combine(directory!.FullName, "Golden");
    }
}

/// <summary>
/// Records every call a decoder makes into its sink as a JSON log: the decode-level contract every client implementation reproduces call for call.
/// </summary>
internal sealed class RecordingSink : ITickSink, ICommandSink
{
    internal JsonArray Log { get; } = [];

    public void BeginTick(uint tick, TickFlags flags, uint periodUs) =>
        Log.Add(new JsonObject { ["call"] = "beginTick", ["tick"] = tick, ["flags"] = (int)flags, ["periodUs"] = periodUs });

    public void BeginEntities(ArchetypePlan archetype) => Log.Add(new JsonObject { ["call"] = "beginEntities", ["archetype"] = archetype.Name });

    public void Enter(uint netId, scoped ReadOnlySpan<double> position, scoped ReadOnlySpan<double> velocity, uint t0, byte epoch) =>
        Log.Add(new JsonObject
        {
            ["call"] = "enter", ["netId"] = netId, ["position"] = Golden.Bits(position), ["velocity"] = Golden.Bits(velocity), ["t0"] = t0, ["epoch"] = epoch,
        });

    public void Segment(uint netId, scoped ReadOnlySpan<double> position, scoped ReadOnlySpan<double> velocity, uint t0, byte epoch) =>
        Log.Add(new JsonObject
        {
            ["call"] = "segment", ["netId"] = netId, ["position"] = Golden.Bits(position), ["velocity"] = Golden.Bits(velocity), ["t0"] = t0, ["epoch"] = epoch,
        });

    public void State(uint netId, byte groupMask) => Log.Add(new JsonObject { ["call"] = "state", ["netId"] = netId, ["groupMask"] = groupMask });

    public void Leave(uint netId) => Log.Add(new JsonObject { ["call"] = "leave", ["netId"] = netId });

    public void Event(MessagePlan type) => Log.Add(new JsonObject { ["call"] = "event", ["type"] = type.Name, ["idx"] = type.Idx });

    public void Self(ArchetypePlan archetype, uint netId, ushort lastSeq, byte ownerMask) =>
        Log.Add(new JsonObject { ["call"] = "self", ["archetype"] = archetype.Name, ["netId"] = netId, ["lastSeq"] = lastSeq, ["ownerMask"] = ownerMask });

    public void Ack(ushort seq, byte reason) => Log.Add(new JsonObject { ["call"] = "ack", ["seq"] = seq, ["reason"] = reason });

    public void Source(ushort requestId, byte status, ushort code) =>
        Log.Add(new JsonObject { ["call"] = "source", ["requestId"] = requestId, ["status"] = status, ["code"] = code });

    public void BeginAggregate(CatalogGrid grid, bool reset) => Log.Add(new JsonObject { ["call"] = "beginAggregate", ["grid"] = grid.Idx, ["reset"] = reset });

    public void AggregateCell(uint cell, scoped ReadOnlySpan<uint> counts)
    {
        var array = new JsonArray();
        foreach (var c in counts)
        {
            array.Add(c);
        }

        Log.Add(new JsonObject { ["call"] = "aggregateCell", ["cell"] = cell, ["counts"] = array });
    }

    public void Metric(MetricPlan metric, int valueIndex, double value) =>
        Log.Add(new JsonObject { ["call"] = "metric", ["name"] = metric.Name, ["index"] = valueIndex, ["value"] = Golden.Bits(value) });

    public void Debug(byte subType, scoped ReadOnlySpan<byte> payload) =>
        Log.Add(new JsonObject { ["call"] = "debug", ["subType"] = subType, ["payload"] = Golden.Hex(payload) });

    public void Ext(uint appTypeId, scoped ReadOnlySpan<byte> payload) =>
        Log.Add(new JsonObject { ["call"] = "ext", ["appTypeId"] = appTypeId, ["payload"] = Golden.Hex(payload) });

    public void UnknownBlock(byte blockType) => Log.Add(new JsonObject { ["call"] = "unknownBlock", ["blockType"] = blockType });

    public void EndTick() => Log.Add(new JsonObject { ["call"] = "endTick" });

    public void Command(MessagePlan type, ushort seq, uint clientTick) =>
        Log.Add(new JsonObject { ["call"] = "command", ["type"] = type.Name, ["idx"] = type.Idx, ["seq"] = seq, ["clientTick"] = clientTick });

    public void Number(FieldPlan field, scoped ReadOnlySpan<double> components) =>
        Log.Add(new JsonObject { ["call"] = "number", ["field"] = field.Name, ["values"] = Golden.Bits(components) });

    public void Text(FieldPlan field, scoped ReadOnlySpan<byte> utf8) =>
        Log.Add(new JsonObject { ["call"] = "text", ["field"] = field.Name, ["utf8"] = Golden.Hex(utf8) });

    public void Bytes(FieldPlan field, scoped ReadOnlySpan<byte> bytes) =>
        Log.Add(new JsonObject { ["call"] = "bytes", ["field"] = field.Name, ["bytes"] = Golden.Hex(bytes) });

    public void List(FieldPlan field, int count, scoped ReadOnlySpan<double> components) =>
        Log.Add(new JsonObject { ["call"] = "list", ["field"] = field.Name, ["count"] = count, ["values"] = Golden.Bits(components) });
}

File: src/Typhon.Client.TypeScript/test/golden-support.ts
import { existsSync, readdirSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  RealmFrame,
  type ArchetypePlan,
  type CommandSink,
  type FieldPlan,
  type GridPlan,
  type MessagePlan,
  type MetricPlan,
  type TickSink,
} from '../src/index.js';

/**
 * Access to the golden vectors the C# reference encoder produced (`test/Typhon.Protocol.Tests/Golden`, 05-sdks § 4),
 * and the recording sink whose log must equal the C# `RecordingSink`'s call for call, key for key.
 */

/**
 * Walks up from this file until the repository's golden directory appears, so the tests run from any working directory.
 */
function findGoldenDirectory(): string {
  let dir = dirname(fileURLToPath(import.meta.url));
  for (;;) {
    const candidate = join(dir, 'test', 'Typhon.Protocol.Tests', 'Golden');
    if (existsSync(candidate)) {
      return candidate;
    }

    const parent = dirname(dir);
    if (parent === dir) {
      throw new Error('test/Typhon.Protocol.Tests/Golden not found above the SDK');
    }

    dir = parent;
  }
}

export const GOLDEN_DIR = findGoldenDirectory();

export function goldenBin(name: string): Uint8Array {
  const buffer = readFileSync(join(GOLDEN_DIR, `${name}.bin`));
  return new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength);
}

/** A vector's expectation; the caller states the shape it reads. */
export function goldenJson(name: string): unknown {
  return JSON.parse(readFileSync(join(GOLDEN_DIR, `${name}.json`), 'utf8'));
}

export function goldenNames(prefix: string): string[] {
  return readdirSync(GOLDEN_DIR)
    .filter((f) => f.startsWith(prefix) && f.endsWith('.json'))
    .map((f) => f.slice(0, -'.json'.length))
    .sort();
}

const bitsView = new DataView(new ArrayBuffer(8));

/**
 * A double's IEEE bits as 16 lower-case hex digits, most significant first — the golden vectors' number format — or
 * `nan`: JavaScript cannot observe a NaN's sign or payload, so the vectors compare NaN as a token.
 */
export function bits(value: number): string {
  if (value !== value) {
    return 'nan';
  }

  bitsView.setFloat64(0, value);
  return bitsView.getUint32(0).toString(16).padStart(8, '0') + bitsView.getUint32(4).toString(16).padStart(8, '0');
}

export function fromBits(hex: string): number {
  if (hex === 'nan') {
    return NaN;
  }

  bitsView.setUint32(0, parseInt(hex.slice(0, 8), 16));
  bitsView.setUint32(4, parseInt(hex.slice(8, 16), 16));
  return bitsView.getFloat64(0);
}

export function bitsOf(values: ArrayLike<number>, count: number): string[] {
  const result: string[] = [];
  for (let i = 0; i < count; i++) {
    result.push(bits(values[i]!));
  }

  return result;
}

/** A realm frame as a golden vector carries it (C# `CatalogSamples.FrameJson`): every number as its IEEE bits. */
export interface FrameJson {
  readonly realmId: number;
  readonly generation: number;
  readonly kindIdx: number;
  readonly appTag: number;
  readonly posBits: number;
  readonly cellM: string;
  readonly deep: boolean;
  readonly min: readonly string[];
  readonly max: readonly string[];
}

export function frameJson(frame: RealmFrame | null): FrameJson | null {
  return frame === null
    ? null
    : {
        realmId: frame.realmId,
        generation: frame.generation,
        kindIdx: frame.kindIdx,
        appTag: frame.appTag,
        posBits: frame.positionBits,
        cellM: bits(frame.cellM),
        deep: frame.deep,
        min: bitsOf(frame.min, 3),
        max: bitsOf(frame.max, 3),
      };
}

export function frameFromJson(json: FrameJson | null | undefined): RealmFrame | null {
  return json == null
    ? null
    : new RealmFrame(
        json.realmId,
        json.generation,
        json.kindIdx,
        json.appTag,
        json.posBits,
        fromBits(json.cellM),
        json.deep,
        json.min.map(fromBits),
        json.max.map(fromBits),
      );
}

export function hex(bytes: Uint8Array): string {
  let s = '';
  for (const b of bytes) {
    s += b.toString(16).padStart(2, '0');
  }

  return s;
}

export function fromHex(text: string): Uint8Array {
  const bytes = new Uint8Array(text.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(text.slice(2 * i, 2 * i + 2), 16);
  }

  return bytes;
}

export type LogEntry = Record<string, unknown>;

const encoder = new TextEncoder();

/** Records every call a decoder makes, in the exact JSON shape of the C# `RecordingSink`. */
export class RecordingSink implements TickSink, CommandSink {
  readonly log: LogEntry[] = [];
  private archetype: ArchetypePlan | null = null;

  beginTick(tick: number, flags: number, periodUs: number): void {
    this.log.push({ call: 'beginTick', tick, flags, periodUs });
  }

  realm(frame: RealmFrame | null): void {
    this.log.push({ call: 'realm', frame: frameJson(frame) });
  }

  beginEntities(archetype: ArchetypePlan): void {
    this.archetype = archetype;
    this.log.push({ call: 'beginEntities', archetype: archetype.name });
  }

  enter(netId: number, position: Float64Array, velocity: Float64Array, t0: Uint32Array, epoch: number): void {
    this.log.push({ call: 'enter', netId, ...this.motion(position, velocity), t0: t0[0], epoch });
  }

  segment(netId: number, position: Float64Array, velocity: Float64Array, t0: Uint32Array, epoch: number): void {
    this.log.push({ call: 'segment', netId, ...this.motion(position, velocity), t0: t0[0], epoch });
  }

  state(netId: number, groupMask: number): void {
    this.log.push({ call: 'state', netId, groupMask });
  }

  leave(netId: number): void {
    this.log.push({ call: 'leave', netId });
  }

  event(type: MessagePlan): void {
    this.log.push({ call: 'event', type: type.name, idx: type.idx });
  }

  self(archetype: ArchetypePlan | null, netId: number, lastSeq: number, ownerMask: number): void {
    this.log.push({ call: 'self', archetype: archetype?.name ?? null, netId, lastSeq, ownerMask });
  }

  ack(seq: number, reason: number): void {
    this.log.push({ call: 'ack', seq, reason });
  }

  source(requestId: number, status: number, code: number): void {
    this.log.push({ call: 'source', requestId, status, code });
  }

  beginAggregate(grid: GridPlan, reset: boolean): void {
    this.log.push({ call: 'beginAggregate', grid: grid.idx, reset });
  }

  aggregateCell(cell: number, counts: Uint32Array): void {
    this.log.push({ call: 'aggregateCell', cell, counts: Array.from(counts) });
  }

  metric(metric: MetricPlan, valueIndex: number, value: number): void {
    this.log.push({ call: 'metric', name: metric.name, index: valueIndex, value: bits(value) });
  }

  debug(subType: number, data: Uint8Array, offset: number, length: number): void {
    this.log.push({ call: 'debug', subType, payload: hex(data.subarray(offset, offset + length)) });
  }

  ext(appTypeId: number, data: Uint8Array, offset: number, length: number): void {
    this.log.push({ call: 'ext', appTypeId, payload: hex(data.subarray(offset, offset + length)) });
  }

  unknownBlock(blockType: number): void {
    this.log.push({ call: 'unknownBlock', blockType });
  }

  endTick(): void {
    this.log.push({ call: 'endTick' });
  }

  command(type: MessagePlan, seq: number, clientTick: number): void {
    this.log.push({ call: 'command', type: type.name, idx: type.idx, seq, clientTick });
  }

  number(field: FieldPlan, values: Float64Array): void {
    this.log.push({ call: 'number', field: field.name, values: bitsOf(values, field.components) });
  }

  text(field: FieldPlan, value: string): void {
    this.log.push({ call: 'text', field: field.name, utf8: hex(encoder.encode(value)) });
  }

  bytes(field: FieldPlan, data: Uint8Array, offset: number, length: number): void {
    this.log.push({ call: 'bytes', field: field.name, bytes: hex(data.subarray(offset, offset + length)) });
  }

  list(field: FieldPlan, count: number, values: Float64Array): void {
    this.log.push({ call: 'list', field: field.name, count, values: bitsOf(values, count * field.components) });
  }

  /** The views as handed out, which the reader sizes exactly: `dims` values, and a velocity only when one travels. */
  private motion(position: Float64Array, velocity: Float64Array): { position: string[]; velocity: string[] } {
    const p = this.archetype?.position ?? null;
    const dims = p === null ? 0 : p.dims;
    if (position.length !== dims || velocity.length !== (p?.vel ? dims : 0)) {
      throw new Error(`motion views of ${position.length} and ${velocity.length} values for ${dims} dimensions`);
    }

    return { position: bitsOf(position, position.length), velocity: bitsOf(velocity, velocity.length) };
  }
}
